File: 1/2020.09.29/111.Minimun Depth of Binary Tree/java/solution.java
class Solution {
    public int minDepth(TreeNode root) {
        // DFS
        if (root == null) {
            return 0;
        }
        int left = minDepth(root.left);
        int right = minDepth(root.right);
        return left == 0 || right == 0 ? left + right + 1 : Math.min(left, right) + 1;
    }

    public int minDepth(TreeNode root) {
        // BFS
        if (root == null) {
            return 0;
        }
        int depth = 0;
        Deque<TreeNode> q = new LinkedList<>();
        q.offerLast(root);
        while (!q.isEmpty()) {
            depth++;
            int size = q.size();
            while (size-- > 0) {
                TreeNode node = q.pollFirst();
                if (node.left == null && node.right == null) {
                    return depth;
                }
                if (node.left != null) {
                    q.offerLast(node.left);
                }
                if (node.right != null) {
                    q.offerLast(node.right);
                }
            }
        }
        return depth;
    }
}

File: 1/2020.09.29/111.Minimun Depth of Binary Tree/python/solution.py
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        # BFS
        depth = 0
        if not root:
            return 0
        q = [root]
        while q:
            n = len(q)
            depth += 1
            for _ in range(n):
                node = q.pop(0)
                if not node.left and not node.right:
                    return depth
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

    def minDepth(self, root: TreeNode) -> int:
        # DFS
        if not root:
            return 0
        if not root.left:
            return self.minDepth(root.right) + 1
        if not root.right:
            return self.minDepth(root.left) + 1
        return min(self.minDepth(root.left), self.minDepth(root.right)) + 1
